server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/EnqueuedMailsDAO.java

```diff
@@ -0,0 +1,291 @@
+import org.apache.mailet.Mail;
+import org.apache.mailet.PerRecipientHeaders;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.UDTValue;
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+class EnqueuedMailsDAO {
+
+    static class EnqueuedMailsDaoUtil {
+
+        static EnqueuedMail toEnqueuedMail(Row row) {
+            MailQueueName queueName = MailQueueName.fromString(row.getString(QUEUE_NAME));
+            Instant timeRangeStart = row.getTimestamp(TIME_RANGE_START).toInstant();
+            BucketId bucketId = BucketId.of(row.getInt(BUCKET_ID));
+            Instant enqueuedTime = row.getTimestamp(ENQUEUED_TIME).toInstant();
+
+            MailAddress sender = Optional.ofNullable(row.getString(SENDER))
+                .map(Throwing.function(MailAddress::new))
+                .orElse(null);
+            List<MailAddress> recipients = row.getList(RECIPIENTS, String.class)
+                .stream()
+                .map(Throwing.function(MailAddress::new))
+                .collect(Guavate.toImmutableList());
+            String state = row.getString(STATE);
+            String remoteAddr = row.getString(REMOTE_ADDR);
+            String remoteHost = row.getString(REMOTE_HOST);
+            String errorMessage = row.getString(ERROR_MESSAGE);
+            String name = row.getString(MAIL_KEY);
+            Date lastUpdated = row.getTimestamp(LAST_UPDATED);
+            Map<String, ByteBuffer> rawAttributes = row.getMap(ATTRIBUTES, String.class, ByteBuffer.class);
+            PerRecipientHeaders perRecipientHeaders = fromHeaderMap(row.getMap(PER_RECIPIENT_SPECIFIC_HEADERS, String.class, UDTValue.class));
+
+            MailImpl mail = MailImpl.builder()
+                .name(name)
+                .sender(sender)
+                .recipients(recipients)
+                .lastUpdated(lastUpdated)
+                .errorMessage(errorMessage)
+                .remoteHost(remoteHost)
+                .remoteAddr(remoteAddr)
+                .state(state)
+                .addAllHeadersForRecipients(perRecipientHeaders)
+                .attributes(toAttributes(rawAttributes))
+                .build();
+
+            return EnqueuedMail.builder()
+                .mail(mail)
+                .bucketId(bucketId)
+                .timeRangeStart(timeRangeStart)
+                .enqueuedTime(enqueuedTime)
+                .mailKey(MailKey.of(name))
+                .mailQueueName(queueName)
+                .build();
+        }
+
+        private static Map<String, Serializable> toAttributes(Map<String, ByteBuffer> rowAttributes) {
+            return rowAttributes.entrySet()
+                .stream()
+                .map(entry -> Pair.of(entry.getKey(), fromByteBuffer(entry.getValue())))
```

Comment:
You might be able to get ride of this intermediate Pair representation by doing these transformation directly when collecting